### PR TITLE
Support retry annotation in functions maven plugin

### PR DESCRIPTION
**Pull request by Flanker32**
Support retry annotation in functions maven plugin, refers https://github.com/Azure/azure-functions-java-library/pull/139

/cc: @amamounelsayed @TsuyoshiUshio @pragnagopa

### CODE REVIEW

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java [hunk 1]
```diff
@@ -84,10 +92,25 @@ public FunctionConfiguration generateConfiguration(final Method method) {
...
+        final ExponentialBackoffRetry exponentialBackoffRetry = method.getDeclaredAnnotation(ExponentialBackoffRetry.class);
+        if (!Objects.isNull(fixedDelayRetry) && !Objects.isNull(exponentialBackoffRetry)) {
+            throw new AzureExecutionException(MULTI_RETRY_ANNOTATION);
+        } else if (fixedDelayRetry != null) {
+            return Retry.createFixedDelayRetryFromAnnotation(fixedDelayRetry);
+        } else if (exponentialBackoffRetry != null) {
+            return Retry.createExponentialBackoffRetryFromAnnotation(exponentialBackoffRetry);
+        } else {
+            return null;
+        }
```
Comment (shenqianjin) [inline azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:111]: q1: line 103 is conflict with line 109/110.

Comment (shenqianjin) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:111]: q2: I guess it should be ok even if both fixedDelayRetry and exponentialBackoffRetry are null. right ?

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/configurations/Retry.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+package com.microsoft.azure.common.function.configurations;
+
+import com.microsoft.azure.functions.annotation.ExponentialBackoffRetry;
+import com.microsoft.azure.functions.annotation.FixedDelayRetry;
+import lombok.Data;
+import lombok.experimental.SuperBuilder;
+
+@SuperBuilder(toBuilder = true)
+@Data
+public class Retry {
```
Comment (shenqianjin) [inline azure-tools-common/src/main/java/com/microsoft/azure/common/function/configurations/Retry.java:16]: are its setXXX() methods used in anywhere? if not, I prefer to user @Getter and @ToString to replace with @Data.

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/configurations/Retry.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public static Retry createExponentialBackoffRetryFromAnnotation(final ExponentialBackoffRetry exponentialBackoffRetry) {
+        return Retry.builder()
+                .strategy(exponentialBackoffRetry.strategy())
+                .maxRetryCount(exponentialBackoffRetry.maxRetryCount())
+                .minimumInterval(exponentialBackoffRetry.minimumInterval())
+                .maximumInterval(exponentialBackoffRetry.maximumInterval()).build();
+    }
+}
```
Comment (shenqianjin) [inline azure-tools-common/src/main/java/com/microsoft/azure/common/function/configurations/Retry.java:37]: how about to just name these two methods as "create" or "build" ?

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java (hunk 1 above)
Comment (Flanker32) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:111]: 103 means exponentialBackoffRetry and fixedDelayRetry could not exists the same time (I'll change to use != to make it more clearly

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java [hunk 4]
```diff
@@ -30,12 +33,17 @@
...
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.function.Function;
 import java.util.function.Supplier;
 
 public class AnnotationHandlerImpl implements AnnotationHandler {
+
+    private static final String MULTI_RETRY_ANNOTATION = "Fixed delay retry and exponential backoff retry is not compatible, " +
```
Comment (andxu) [inline azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:44]: is -> are

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java [hunk 5]
```diff
@@ -84,10 +92,25 @@ public FunctionConfiguration generateConfiguration(final Method method) {
...
+        config.setRetry(getRetryConfigurationFromMethod(method));
         config.setEntryPoint(method.getDeclaringClass().getCanonicalName() + "." + method.getName());
         return config;
     }
 
+    private Retry getRetryConfigurationFromMethod(Method method) throws AzureExecutionException {
+        final FixedDelayRetry fixedDelayRetry = method.getDeclaredAnnotation(FixedDelayRetry.class);
+        final ExponentialBackoffRetry exponentialBackoffRetry = method.getDeclaredAnnotation(ExponentialBackoffRetry.class);
+        if (!Objects.isNull(fixedDelayRetry) && !Objects.isNull(exponentialBackoffRetry)) {
+            throw new AzureExecutionException(MULTI_RETRY_ANNOTATION);
```
Comment (andxu) [inline azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:104]: remove else

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/configurations/Retry.java (hunk 3 above)
Comment (Flanker32) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/configurations/Retry.java:37]: I think its okay to keep the name to indicate its strategy type

Comment (shenqianjin) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/configurations/Retry.java:37]: yes.

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java (hunk 5 above)
Comment (Flanker32) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:104]: No need to evaluate conditions below in this method, so if else may be better?

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java [hunk 6]
```diff
@@ -84,10 +92,25 @@ public FunctionConfiguration generateConfiguration(final Method method) {
...
+        final ExponentialBackoffRetry exponentialBackoffRetry = method.getDeclaredAnnotation(ExponentialBackoffRetry.class);
+        if (fixedDelayRetry != null && exponentialBackoffRetry != null) {
+            throw new AzureExecutionException(MULTI_RETRY_ANNOTATION);
+        } else if (fixedDelayRetry != null) {
+            return Retry.createFixedDelayRetryFromAnnotation(fixedDelayRetry);
+        } else if (exponentialBackoffRetry != null) {
+            return Retry.createExponentialBackoffRetryFromAnnotation(exponentialBackoffRetry);
+        } else {
+            return null;
+        }
```
Comment (shenqianjin) [inline azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:111]: Iprefer to:

if (fixedDelayRetry != null) {
            return Retry.createFixedDelayRetryFromAnnotation(fixedDelayRetry);
 }
if (exponentialBackoffRetry != null) {
            return Retry.createExponentialBackoffRetryFromAnnotation(exponentialBackoffRetry);
}
return null;

It's up to you.

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java (hunk 5 above)
Comment (andxu) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:104]: the behaviors are same, but is more readable:https://stackoverflow.com/questions/2681883/unnecessary-else-statement

Comment (Flanker32) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:104]: Fixed

File: azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java (hunk 6 above)
Comment (Flanker32) [reply azure-tools-common/src/main/java/com/microsoft/azure/common/function/handlers/AnnotationHandlerImpl.java:111]: Fixed

**Review by shenqianjin — APPROVED** [review]
LGTM

**Pull request merged**